Enable ssm to send command output to cloudwatch logs

bernard-sh opened the pull request:
Hi, aws just enabled that ssm to send the output to cloudwatch logs, this diff is to enabled ssm-sh to use this feature, please review.

Commit:
Enable ssm to send command output to cloudwatch logs

Commit:
Enable ssm to send command output to cloudwatch logs

Commit:
Enable ssm to send command output to cloudwatch logs

itsdalmo commented:
Thanks for the PR; the feature and code looks great :tada: Bit busy this weekend unfortunately, but will try to make time to review it! One thing - could you vendor the cloudwatch module so we can see if travis turns green?

itsdalmo commented on manager/manager.go:
```diff
@@ -314,13 +330,22 @@ func (m *Manager) newCommandOutput(result *ssm.GetCommandInvocationOutput, err e
 	case "Success":
 		out.Output = aws.StringValue(result.StandardOutputContent)
 		out.OutputUrl = aws.StringValue(result.StandardOutputUrl)
+		out.CwlGroup = aws.StringValue(result.CloudWatchOutputConfig.CloudWatchLogGroupName)
```
I think we can safely use `m.logGroupName` here. As a user, if SSM suddenly changed the log group used - I'd rather see an error than output + no error. What do you think?

bernard-sh replied:
So you mean get the log group name from `m.logGroupName` instead of from `result`?

itsdalmo replied:
Yeah, that was my thought. What do you think?

bernard-sh replied:
If that is the case, i think we should also notify the user, that SSM just use a different logs that specify by user. What do you think?

itsdalmo replied:
After more deliberation, I think it's better to do as you have done 👍 No change needed!

itsdalmo commented on manager/manager.go:
```diff
@@ -334,12 +359,21 @@ func (m *Manager) newCommandOutput(result *ssm.GetCommandInvocationOutput, err e
+
+			cwlout, err := m.readCwl(out.CwlGroup, out.CwlStream)
+			if err != nil {
+				out.Error = errors.Wrap(err, "failed to fetch extended output from cloudwatchlogs")
```
Nitpick; Missing space between "cloudwatch" and  "logs"

itsdalmo commented on manager/manager.go:
```diff
@@ -376,3 +410,33 @@ func (m *Manager) readS3Object(bucket, key string) (string, error) {
+	if len(cwls.LogStreams) > 1 {
+		return "", errors.Errorf("Find multiple logstream for %s - %s", commandId, instanceId)
+	}
+	return aws.StringValue(cwls.LogStreams[0].LogStreamName), nil
```
I think we need to check for zero-length/nil output as well when getting the first index of an array, currently it panics (as I discovered when I tried to output to a log group that the instance did not have permissions to log to).

itsdalmo commented on manager/manager.go:
```diff
@@ -376,3 +410,33 @@ func (m *Manager) readS3Object(bucket, key string) (string, error) {
+		return "", err
+	}
+	if len(cwls.LogStreams) > 1 {
+		return "", errors.Errorf("Find multiple logstream for %s - %s", commandId, instanceId)
```
Nitpick "found" instead of "Find" (no capital letters either).

itsdalmo reviewed (commented):
Had a quick look now (will take a closer look as soon as I have time) and left some comments! Also worth mentioning that it would be great if we could add a couple of test cases to avoid breaking this functionality in the future - if you are up for it, we'd need test cases like:
- log group set _with_ extended output.
- log group set _without_ extended output.

And I noticed that [PrintCommandOutput](https://github.com/SiahaanBernard/ssm-sh/blob/master/command/utils.go#L94) does not know what to do with the extra information passed via [CommandOutput](https://github.com/SiahaanBernard/ssm-sh/blob/Enable_cwl_output/manager/manager.go#L46-L47). I think we should probably print the log group name and log stream?

Thanks again for the PR! 🙌

Commit:
Update vendor

Commit:
Update vendor

Commit:
Update vendor

Commit:
Update vendor

Commit:
remove dep ensure command from travis.yml

Commit:
remove dep ensure command from travis.yml

Commit:
remove dep ensure command from travis.yml

Commit:
remove dep ensure command from travis.yml

geowa4 committed:
Fix certificate error in Docker image.

When running `list instances`, I would receive an error `Post
https://ssm.us-west-2.amazonaws.com/: x509: failed to load system roots and no
roots provided`. Adding `ca-certificates` into the image resolves this issue.

I also reordered the directives for the final image to reuse layers. The layers
are minimal so you won't notice, but it's a good habit.

Commit:
Merge pull request #22 from geowa4/fix-certificate-error

Fix certificate error in Docker image.

Commit:
- Fix typos on error messages
- Print cloudwatch log group and log stream name, when print the output
- Check the length of the output on cloudwatch stream

Commit:
- Fix typos on error messages
- Print cloudwatch log group and log stream name, when print the output
- Check the length of the output on cloudwatch stream

Commit:
- Fix typos on error messages
- Print cloudwatch log group and log stream name, when print the output
- Check the length of the output on cloudwatch stream

itsdalmo commented:
Hey! Sorry for the radio silence on my end; what is the status of this PR? I'd really like to see this merged, do you think you'll have time finish it?

Commit:
Add CACHE_TAG argument to dockerfile. Use alpine latest explicitly.

Commit:
Slight makefile cleanup to support CACHE_TAG on dockerhub.

Commit:
Remove deprecated MAINTAINER from dockerfile.

Commit:
Add ENV variable for CACHE_TAG.

Commit:
Remove ENV entry for CACHE_TAG until we can test with tags.

Commit:
Merge pull request #23 from itsdalmo/fix-version-in-docker

Fix version in docker

Commit:
Revert "Fix version in docker"

Commit:
Retain the makefile changes.

Commit:
Retain Dockerfile changes.

Commit:
Merge pull request #24 from itsdalmo/revert-23-fix-version-in-docker

Revert "Fix version in docker"

bernard-sh commented:
I'l try to finish this weekend ya.

Commit:
Check error when cannot describe log stream output.

Commit:
Check error when cannot describe log stream output.

Commit:
Check error when cannot describe log stream output.

Commit:
Check error when cannot describe log stream output.

Commit:
remove unused file

Commit:
remove unused file

Commit:
remove unused file

Commit:
remove unused file

Commit:
Update terraform example. (#25)

- Updated terraform/provider/module dependencies and cleaned up the code.
- Example now uses default VPC/subnets and latest Amazon Linux 2 AMI.
- Example now generates the output bucket and outputs the name.
- Also generating a log group for testing CloudWatch output (for #21).

itsdalmo commented:
Ref this commit: https://github.com/itsdalmo/ssm-sh/commit/2a2860692d7ce2680d6ae458a87fd5c37703a9ed

I have updated the terraform example to generate a CloudWatch logs group and grant the instance permissions to write to the log group, but on building the latest commit in your branch and testing the `ssm-sh` binary I get the following error:

```bash
$ ./ssm-sh run command -t <instance-id> -l  ssm-sh-example-output-<account-id> -x -- ls -la

<instance-id> - Success:
can't find log stream with prefix b6be0895-2dbd-430d-9a4f-2147a332bedc/i-00e303a0eba0de85d in ssm-sh-example-output-<account-id> log group

<rest of output>
```

Perhaps I got the permissions wrong for CloudWatch output? You can see them here:
https://github.com/itsdalmo/ssm-sh/blob/master/terraform/modules/ssm-example/main.tf#L110-L123

Commit:
Merge branch 'Enable_cwl_output' of github.com:SiahaanBernard/ssm-sh into Enable_cwl_output

Commit:
Update terraform/example.tf to add managed policy to access cloudwatch logs

Commit:
Update terraform/example.tf to add managed policy to access cloudwatch logs

Commit:
Update terraform/README.md to give explanation about ssm-cwl-policy

Commit:
Update terraform/README.md to give explanation about ssm-cwl-policy

bernard-sh commented:
Hi sorry forget to mention it, user or roles that executed ssm-sh need to have permission `logs:DescribeLogStreams` and `logs:GetLogEvents` to ssm run command output loggroups, or *, so that ssm-sh can get the log streams name, and getEvent from it. I update the terraform example to create managed policy to allow access to Cloudwatch Logs.

itsdalmo commented:
The diff looks a bit iffy after you have merged the commits into your `Enable_cwl_output` branch; I think perhaps you need to merge the latest commits from upstream `master` into the `master` branch of your fork - and then perhaps rebase the `Enable_cwl_output` branch? (Disclaimer; I'm not a git wizard).

Also, I'm not sure the changes you made to the terraform solves the problem I had. The role I was testing with already has admin privileges, so attaching the managed policy should not make a difference? 

Based on the error message, it finds a log stream in the output, but that log stream never shows up in Cloudwatch, which suggests to me that it is the _instance_ that lacks the necessary privileges to create the log stream in the first place?

Commit:
Merge branch 'Enable_cwl_output' of github.com:SiahaanBernard/ssm-sh into Enable_cwl_output

bernard-sh commented:
Hi
Regarding the diff, this is actually my first time doing PR, thanks for be nice, already fixed it.
Regarding CW Logs stream, my bad, i don't read the error message carefully.
I'm pretty sure, policy you created and attached to ec2 already sufficient. 
I found different behaviour when i try this with different AMI, but same policy. I try to apply the terraform and run ssm-sh run command, i found issue like yours. I try to look to error message on instance, and cannot find ssm error logs file on `/var/log/amazon/ssm`, but when i try to change the AMI on terraform config using AMI (ubuntu), the logstream is created, and  the error log is created. Will continue investigation later.

itsdalmo commented:
Great stuff @SiahaanBernard - I'll try to help with figuring out the correct IAM permissions when I have time 👍

itsdalmo commented:
Sorry I dropped the ball on this one. I'll close the PR now since it's stale, and also the AWS CLI has session manager baked in.
